File: packages/typescript/core/src/scaling.ts
/**
 * Tree scaling utilities: depth truncation, node-budget compaction,
 * salience filtering, and subtree extraction.
 *
 * These operate on wire-format SlopNodes and are used by both the
 * client (in-browser provider) and server (multi-connection provider)
 * to respect consumer token budgets.
 */

import type { SlopNode } from "./types";

/** Options for preparing a tree for output to a consumer. */
export interface OutputTreeOptions {
  /** Maximum depth to resolve. Nodes beyond this become stubs with summaries. */
  maxDepth?: number;
  /** Maximum total nodes. Lowest-salience subtrees are collapsed first. */
  maxNodes?: number;
  /** Minimum salience threshold. Nodes below this are excluded. */
  minSalience?: number;
  /** Only include nodes of these types. */
  types?: string[];
}

/**
 * Prepare a tree for output to a consumer by applying depth truncation,
 * salience filtering, type filtering, and node-budget compaction.
 */
export function prepareTree(root: SlopNode, options: OutputTreeOptions): SlopNode {
  let tree = root;
  if (options.minSalience != null || options.types != null) {
    tree = filterTree(tree, options.minSalience, options.types);
  }
  if (options.maxDepth != null) {
    tree = truncateTree(tree, options.maxDepth);
  }
  if (options.maxNodes != null) {
    tree = autoCompact(tree, options.maxNodes);
  }
  return tree;
}

/**
 * Extract a subtree rooted at the given path.
 * Path is slash-separated node IDs: "/inbox/msg-42".
 * Returns undefined if the path doesn't exist.
 */
export function getSubtree(root: SlopNode, path: string): SlopNode | undefined {
  if (!path || path === "/") return root;

  const segments = path.replace(/^\//, "").split("/").filter(Boolean);
  let current = root;

  for (const seg of segments) {
    const child = current.children?.find((c) => c.id === seg);
    if (!child) return undefined;
    current = child;
  }

  return current;
}

/**
 * Truncate a tree at the given depth. Nodes beyond the depth become stubs
 * with `meta.total_children` and `meta.summary`.
 */
export function truncateTree(node: SlopNode, depth: number): SlopNode {
  if (depth <= 0 && node.children?.length) {
    return {
      id: node.id,
      type: node.type,
      ...(node.properties && { properties: node.properties }),
      meta: {
        ...node.meta,
        total_children: node.children.length,
      },
    };
  }
  if (!node.children) return node;
  return {
    ...node,
    children: node.children.map((c) => truncateTree(c, depth - 1)),
  };
}

/**
 * Compact a tree to fit within a node budget. Collapses lowest-salience
 * subtrees first, preserving root children and pinned nodes.
 */
export function autoCompact(root: SlopNode, maxNodes: number): SlopNode {
  const total = countNodes(root);
  if (total <= maxNodes) return root;

  const candidates: CompactCandidate[] = [];
  if (root.children) {
    for (let i = 0; i < root.children.length; i++) {
      collectCandidates(root.children[i], [i], candidates, false);
    }
  }

  candidates.sort((a, b) => a.score - b.score);

  const tree = structuredClone(root);
  let nodeCount = total;

  for (const candidate of candidates) {
    if (nodeCount <= maxNodes) break;
    const saved = collapseAtPath(tree, candidate.path);
    nodeCount -= saved;
  }

  return tree;
}

/**
 * Filter a tree by salience threshold and/or node types.
 * Nodes below minSalience or not matching types are removed.
 * The root node is never filtered.
 */
export function filterTree(
  node: SlopNode,
  minSalience?: number,
  types?: string[]
): SlopNode {
  if (!node.children) return node;

  const filtered = node.children
    .filter((child) => {
      if (minSalience != null) {
        const salience = child.meta?.salience ?? 0.5;
        if (salience < minSalience) return false;
      }
      if (types != null && !types.includes(child.type)) return false;
      return true;
    })
    .map((child) => filterTree(child, minSalience, types));

  return {
    ...node,
    children: filtered.length > 0 ? filtered : undefined,
  };
}

/** Count total nodes in a tree. */
export function countNodes(node: SlopNode): number {
  return (
    1 + (node.children?.reduce((sum, c) => sum + countNodes(c), 0) ?? 0)
  );
}

// --- Internal helpers ---

interface CompactCandidate {
  path: number[];
  score: number;
  childCount: number;
}

function collectCandidates(
  node: SlopNode,
  path: number[],
  candidates: CompactCandidate[],
  isRootChild: boolean = false
): void {
  if (!node.children) return;
  for (let i = 0; i < node.children.length; i++) {
    const child = node.children[i];
    const childPath = [...path, i];

    if (child.children?.length && !isRootChild && !child.meta?.pinned) {
      const childCount = countNodes(child) - 1;
      const salience = child.meta?.salience ?? 0.5;
      const depth = childPath.length;
      const score = salience - depth * 0.01 - childCount * 0.001;
      candidates.push({ path: childPath, score, childCount });
    }

    collectCandidates(child, childPath, candidates, false);
  }
}

function collapseAtPath(tree: SlopNode, path: number[]): number {
  let node = tree;
  for (let i = 0; i < path.length - 1; i++) {
    if (!node.children?.[path[i]]) return 0;
    node = node.children[path[i]];
  }

  const idx = path[path.length - 1];
  if (!node.children?.[idx]) return 0;

  const target = node.children[idx];
  const saved = countNodes(target) - 1;

  node.children[idx] = {
    id: target.id,
    type: target.type,
    ...(target.properties && { properties: target.properties }),
    ...(target.affordances && { affordances: target.affordances }),
    meta: {
      ...target.meta,
      total_children: target.children?.length ?? 0,
      summary:
        target.meta?.summary ?? `${target.children?.length ?? 0} children`,
    },
  };

  return saved;
}

File: packages/python/slop-ai/src/slop_ai/scaling.py
"""Tree scaling utilities: depth truncation, node-budget compaction,
salience filtering, and subtree extraction.

These operate on wire-format SlopNodes and are used by providers
to respect consumer token budgets.
"""

from __future__ import annotations

import copy
from dataclasses import dataclass
from typing import Any

from .types import NodeMeta, SlopNode


@dataclass
class OutputTreeOptions:
    """Options for preparing a tree for output to a consumer."""

    max_depth: int | None = None
    max_nodes: int | None = None
    min_salience: float | None = None
    types: list[str] | None = None


def prepare_tree(root: SlopNode, options: OutputTreeOptions) -> SlopNode:
    """Prepare a tree for output by applying filter → truncate → compact."""
    tree = root
    if options.min_salience is not None or options.types is not None:
        tree = filter_tree(tree, options.min_salience, options.types)
    if options.max_depth is not None:
        tree = truncate_tree(tree, options.max_depth)
    if options.max_nodes is not None:
        tree = auto_compact(tree, options.max_nodes)
    return tree


def get_subtree(root: SlopNode, path: str) -> SlopNode | None:
    """Extract a subtree by slash-separated node ID path (e.g. '/inbox/msg-42')."""
    if not path or path == "/":
        return root

    segments = [s for s in path.strip("/").split("/") if s]
    current = root
    for seg in segments:
        if not current.children:
            return None
        found = None
        for child in current.children:
            if child.id == seg:
                found = child
                break
        if found is None:
            return None
        current = found
    return current


def truncate_tree(node: SlopNode, depth: int) -> SlopNode:
    """Collapse nodes beyond depth to stubs with meta.total_children."""
    if depth <= 0 and node.children:
        meta = copy.copy(node.meta) if node.meta else NodeMeta()
        meta.total_children = len(node.children)
        return SlopNode(
            id=node.id,
            type=node.type,
            properties=node.properties,
            meta=meta,
        )
    if not node.children:
        return node
    return SlopNode(
        id=node.id,
        type=node.type,
        properties=node.properties,
        children=[truncate_tree(c, depth - 1) for c in node.children],
        affordances=node.affordances,
        meta=node.meta,
    )


def auto_compact(root: SlopNode, max_nodes: int) -> SlopNode:
    """Collapse lowest-salience subtrees to fit within a node budget.

    Preserves root children and pinned nodes.
    """
    total = count_nodes(root)
    if total <= max_nodes:
        return root

    candidates: list[_CompactCandidate] = []
    if root.children:
        for i, child in enumerate(root.children):
            _collect_candidates(child, [i], candidates, is_root_child=False)

    candidates.sort(key=lambda c: c.score)

    tree = copy.deepcopy(root)
    node_count = total

    for candidate in candidates:
        if node_count <= max_nodes:
            break
        saved = _collapse_at_path(tree, candidate.path)
        node_count -= saved

    return tree


def filter_tree(
    node: SlopNode,
    min_salience: float | None = None,
    types: list[str] | None = None,
) -> SlopNode:
    """Filter a tree by salience threshold and/or node types.

    The root node is never filtered.
    """
    if not node.children:
        return node

    filtered: list[SlopNode] = []
    for child in node.children:
        if min_salience is not None:
            salience = child.meta.salience if child.meta and child.meta.salience is not None else 0.5
            if salience < min_salience:
                continue
        if types is not None and child.type not in types:
            continue
        filtered.append(filter_tree(child, min_salience, types))

    return SlopNode(
        id=node.id,
        type=node.type,
        properties=node.properties,
        children=filtered if filtered else None,
        affordances=node.affordances,
        meta=node.meta,
    )


def count_nodes(node: SlopNode) -> int:
    """Count total nodes in a tree."""
    return 1 + sum(count_nodes(c) for c in (node.children or []))


# --- Internal helpers ---


@dataclass
class _CompactCandidate:
    path: list[int]
    score: float
    child_count: int


def _collect_candidates(
    node: SlopNode,
    path: list[int],
    candidates: list[_CompactCandidate],
    is_root_child: bool = False,
) -> None:
    if not node.children:
        return
    for i, child in enumerate(node.children):
        child_path = path + [i]

        if (
            child.children
            and not is_root_child
            and not (child.meta and child.meta.pinned)
        ):
            child_count = count_nodes(child) - 1
            salience = child.meta.salience if child.meta and child.meta.salience is not None else 0.5
            depth = len(child_path)
            score = salience - depth * 0.01 - child_count * 0.001
            candidates.append(_CompactCandidate(path=child_path, score=score, child_count=child_count))

        _collect_candidates(child, child_path, candidates, is_root_child=False)


def _collapse_at_path(tree: SlopNode, path: list[int]) -> int:
    node = tree
    for i in range(len(path) - 1):
        if not node.children or path[i] >= len(node.children):
            return 0
        node = node.children[path[i]]

    idx = path[-1]
    if not node.children or idx >= len(node.children):
        return 0

    target = node.children[idx]
    saved = count_nodes(target) - 1

    meta = copy.copy(target.meta) if target.meta else NodeMeta()
    meta.total_children = len(target.children) if target.children else 0
    if not meta.summary:
        meta.summary = f"{len(target.children) if target.children else 0} children"

    node.children[idx] = SlopNode(
        id=target.id,
        type=target.type,
        properties=target.properties,
        affordances=target.affordances,
        meta=meta,
    )

    return saved
